### PR TITLE
Add `-Zmiri-env-set` to set environment variables without modifying the host environment

### PR DESCRIPTION
**Pull request by eduardosm**
This option allows to pass environment variables to the interpreted program without needing to modify the host environment (which may have undesired effects in some cases).

### CODE REVIEW

**Comment by RalfJung** [comment]
Doesn't rustc have some flag to sent env vars nowadays? Should we just reuse that instead of adding our own?

**Comment by eduardosm** [comment]
I'm not aware of such flag

**Comment by RalfJung** [comment]
https://github.com/rust-lang/compiler-team/issues/653 was the decision to add it. Tracking issue is at https://github.com/rust-lang/rust/issues/118372, so it seems to have been implemented.

**Comment by eduardosm** [comment]
Do we want the same flag for compile-time `env!` and run-time `std::env::var`?

File: README.md [hunk 1]
```diff
@@ -311,6 +311,10 @@ environment variable. We first document the most relevant and most commonly used
 * `-Zmiri-env-forward=<var>` forwards the `var` environment variable to the interpreted program. Can
   be used multiple times to forward several variables. Execution will still be deterministic if the
   value of forwarded variables stays the same. Has no effect if `-Zmiri-disable-isolation` is set.
+* `-Zmiri-env-set=<var>=<value>` sets the `var` environment variable to `value` in the interpreted.
+  It can be used to pass environment variables without needing to alter the host environment. It can
+  be used multiple times to set several variables. If `-Zmiri-disable-isolation` is set, values set
```
Comment (RalfJung) [inline README.md:316]:
```suggestion
  be used multiple times to set several variables. If `-Zmiri-disable-isolation` or `-Zmiri-env-forward` is set, values set
```

File: tests/pass/shims/env/var-set.rs [hunk 2]
```diff
@@ -0,0 +1,7 @@
+// Test both a value set and unset in the host.
```
Comment (RalfJung) [inline tests/pass/shims/env/var-set.rs:1]:
```suggestion
// Test a value set on the host (MIRI_ENV_VAR_TEST) and one that is not.
```

**Comment by RalfJung** [comment]
> Do we want the same flag for compile-time `env!` and run-time `std::env::var`?

Fair, I guess it makes sense to treat them separately.

File: README.md [hunk 3]
```diff
@@ -311,6 +311,10 @@ environment variable. We first document the most relevant and most commonly used
 * `-Zmiri-env-forward=<var>` forwards the `var` environment variable to the interpreted program. Can
   be used multiple times to forward several variables. Execution will still be deterministic if the
   value of forwarded variables stays the same. Has no effect if `-Zmiri-disable-isolation` is set.
+* `-Zmiri-env-set=<var>=<value>` sets the `var` environment variable to `value` in the interpreted.
```
Comment (oli-obk) [inline README.md:314]:
```suggestion
* `-Zmiri-env-set=<var>=<value>` sets the `var` environment variable to `value` in the interpreted program.
```

**Review by oli-obk — CHANGES_REQUESTED** [review]

**Review by oli-obk — APPROVED** [review]

**Pull request merged**